### PR TITLE
Enable and constrain reuse for GitVersion installations

### PR DESCRIPTION
**Pull request by psakievich**
## TLDR: 

This PR fixes two issues:

1) Spack's concretizer failed when a `GitVersion` had a prior install with the same version
2) Fixing 1) exposed that the concretizer's re-use would pick a `GitVersion` over `StandardVersions` if they were installed sooner and met the `satisfies` criteria. We want it to only pick `GitVersions` when the version is exactly equal.

--------------------------------
## Original Post

Currently the concretizer fails if you reuse a git ref version that has already been installed but modify the spec at all.

See #38484 for futher diagnosis

The issue here is that since there is no established provenance for these versions the highest weight they are currently assigned is that of prior install. Re-use checks then fail because the weight of the version is identical to the solver.

Ironically, these versions are given the highest weights possible when specified on the CLI for the first time.  They should only appear in a DAG if they are an exact match or if the user specifies them at the CLI. Therefore it makes sense to preserve their higher ordering.

Getting this right is critical to moving all branch based versions to a pinned git-ref in the future.

### CODE REVIEW

**Comment by psakievich** [comment]
@spackbot fix style

**Comment by psakievich** [comment]
Interestingly this PR is currently giving the wrong reuse results for a standard case though:
``` console
$ spack solve zlib-ng
==> Best of 4 considered solutions.
==> Optimization Criteria:
  Priority  Criterion                                            Installed  ToBuild
  1         number of packages to build (vs. reuse)                      -        0
  2         number of nodes from the same package                        -        0
  3         requirement weight                                           0        0
  4         deprecated versions used                                     0        0
  5         version weight                                               5        0
  6         number of non-default variants (roots)                       0        0
  7         preferred providers for roots                                0        0
  8         default values of variants not being used (roots)            0        0
  9         number of non-default variants (non-roots)                   0        0
  10        preferred providers (non-roots)                              0        0
  11        compiler mismatches that are not from CLI                    0        0
  12        compiler mismatches that are not from CLI                    0        0
  13        non-preferred OS's                                           0        0
  14        version badness                                              5        0
  15        default values of variants not being used (non-roots)        0        0
  16        non-preferred compilers                                      0        0
  17        target mismatches                                            0        0
  18        non-preferred targets                                        0        0
  19        compiler mismatches (runtimes)                               0        0
  20        version badness (runtimes)                                   0        0
  21        non-preferred targets (runtimes)                             0        0
  22        edge wiring                                                  0        0

[+]  zlib-ng@git.2.1.5=2.1.5%apple-clang@15.0.0+compat+new_strategies+opt+pic+shared build_system=autotools arch=darwin-sonoma-m1
[+]      ^gmake@4.4.1%apple-clang@15.0.0~guile build_system=generic arch=darwin-sonoma-m1
[+]      ^gnuconfig@2022-09-17%apple-clang@15.0.0 build_system=generic arch=darwin-sonoma-m1

$ spack find zlib-ng
-- darwin-sonoma-m1 / apple-clang@15.0.0 ------------------------
zlib-ng@2.1.5  zlib-ng@git.2.1.5=2.1.5
==> 2 installed packages
```
Re-use should be picking up the non-git ref install.  According to how we have written `satisfies` for `GitRefVersions` and `StandardVersions`, and my memory, it should be impossible for a `GitRefVersion` to get picked. Or at least that was the goal. Will need to dig deeper into this before this can be merged.

File: lib/spack/spack/solver/asp.py [hunk 1]
```diff
@@ -2000,7 +1999,10 @@ def define_ad_hoc_versions_from_specs(
             # best possible, so they're guaranteed to be used preferentially.
             version = s.versions.concrete
 
-            if version is None or any(v == version for v in self.possible_versions[s.name]):
+            if version is None or any(
+                (v == version and not isinstance(version, vn.GitVersion))
+                for v in self.possible_versions[s.name]
+            ):
```
Comment (becker33) [inline lib/spack/spack/solver/asp.py]: This does extra comparisons relative to:

```suggestion
            if version is None or (not isinstance(version, vn.GitVersion) and any(
                v == version
                for v in self.possible_versions[s.name]
            )):
```

**Comment by psakievich** [comment]
I"m not so sure the issue with reuse I outlined is an issue.  Basically, spack is picking the latest install that meets the spec requirements.  If I do:
``` console
$ spack install zlib-ng@3f35bfccff2d1dacdfe9844712be1e042d028700=2.1.5
[+] /Users/psakiev/scratch/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/gmake-4.4.1-yg5a2ohrgzyd532q2wnw5agfpzqgqfsb
[+] /Users/psakiev/scratch/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/gnuconfig-2022-09-17-7gytrfuudfw2tpponhl5i4x4xtmbb6wc
==> Installing zlib-ng-3f35bfccff2d1dacdfe9844712be1e042d028700=2.1.5-q4dw5qjorcvw7bwp7svflr6fnmd6e2lg [3/3]
==> No binary for zlib-ng-3f35bfccff2d1dacdfe9844712be1e042d028700=2.1.5-q4dw5qjorcvw7bwp7svflr6fnmd6e2lg found: installing from source
==> No patches needed for zlib-ng
==> zlib-ng: Executing phase: 'autoreconf'
==> zlib-ng: Executing phase: 'configure'
==> zlib-ng: Executing phase: 'build'
==> zlib-ng: Executing phase: 'install'
==> zlib-ng: Successfully installed zlib-ng-3f35bfccff2d1dacdfe9844712be1e042d028700=2.1.5-q4dw5qjorcvw7bwp7svflr6fnmd6e2lg
  Stage: 10.90s.  Autoreconf: 0.38s.  Configure: 8.93s.  Build: 1.59s.  Install: 0.18s.  Post-install: 0.05s.  Total: 22.18s
[+] /Users/psakiev/scratch/spack/opt/spack/darwin-sonoma-m1/apple-clang-15.0.0/zlib-ng-3f35bfccff2d1dacdfe9844712be1e042d028700_2.1.5-q4dw5qjorcvw7bwp7svflr6fnmd6e2lg
;As1082338:spack psakiev$ ng3f35bfccff2d1dacd@3f35bfccff2d1dacdfe9844712be1e042d028700=2.1.5

$ spack solve zlib-ng
==> Best of 5 considered solutions.
==> Optimization Criteria:
  Priority  Criterion                                            Installed  ToBuild
  1         number of packages to build (vs. reuse)                      -        0
  2         number of nodes from the same package                        -        0
  3         requirement weight                                           0        0
  4         deprecated versions used                                     0        0
  5         version weight                                               5        0
  6         number of non-default variants (roots)                       0        0
  7         preferred providers for roots                                0        0
  8         default values of variants not being used (roots)            0        0
  9         number of non-default variants (non-roots)                   0        0
  10        preferred providers (non-roots)                              0        0
  11        compiler mismatches that are not from CLI                    0        0
  12        compiler mismatches that are not from CLI                    0        0
  13        non-preferred OS's                                           0        0
  14        version badness                                              5        0
  15        default values of variants not being used (non-roots)        0        0
  16        non-preferred compilers                                      0        0
  17        target mismatches                                            0        0
  18        non-preferred targets                                        0        0
  19        compiler mismatches (runtimes)                               0        0
  20        version badness (runtimes)                                   0        0
  21        non-preferred targets (runtimes)                             0        0
  22        edge wiring                                                  0        0

[+]  zlib-ng@3f35bfccff2d1dacdfe9844712be1e042d028700=2.1.5%apple-clang@15.0.0+compat+new_strategies+opt+pic+shared build_system=autotools arch=darwin-sonoma-m1
[+]      ^gmake@4.4.1%apple-clang@15.0.0~guile build_system=generic arch=darwin-sonoma-m1
[+]      ^gnuconfig@2022-09-17%apple-clang@15.0.0 build_system=generic arch=darwin-sonoma-m1
```
It is picking the latest install, but to be fair to the concretizer, it is just the latest spec installed that meets the criteria I gave.

This one I think could be wrong, given a `StandardVersion` install already exists?
``` console
$ spack solve zlib-ng@2
==> Best of 4 considered solutions.
==> Optimization Criteria:
  Priority  Criterion                                            Installed  ToBuild
  1         number of packages to build (vs. reuse)                      -        0
  2         number of nodes from the same package                        -        0
  3         requirement weight                                           0        0
  4         deprecated versions used                                     0        0
  5         version weight                                               5        0
  6         number of non-default variants (roots)                       0        0
  7         preferred providers for roots                                0        0
  8         default values of variants not being used (roots)            0        0
  9         number of non-default variants (non-roots)                   0        0
  10        preferred providers (non-roots)                              0        0
  11        compiler mismatches that are not from CLI                    0        0
  12        compiler mismatches that are not from CLI                    0        0
  13        non-preferred OS's                                           0        0
  14        version badness                                              5        0
  15        default values of variants not being used (non-roots)        0        0
  16        non-preferred compilers                                      0        0
  17        target mismatches                                            0        0
  18        non-preferred targets                                        0        0
  19        compiler mismatches (runtimes)                               0        0
  20        version badness (runtimes)                                   0        0
  21        non-preferred targets (runtimes)                             0        0
  22        edge wiring                                                  0        0

[+]  zlib-ng@3f35bfccff2d1dacdfe9844712be1e042d028700=2.1.5%apple-clang@15.0.0+compat+new_strategies+opt+pic+shared build_system=autotools arch=darwin-sonoma-m1
[+]      ^gmake@4.4.1%apple-clang@15.0.0~guile build_system=generic arch=darwin-sonoma-m1
[+]      ^gnuconfig@2022-09-17%apple-clang@15.0.0 build_system=generic arch=darwin-sonoma-m1
```

This one is most certainly wrong since `@2.1.5` is installed:
``` console
$ spack solve zlib-ng@2.1.5
==> Best of 3 considered solutions.
==> Optimization Criteria:
  Priority  Criterion                                            Installed  ToBuild
  1         number of packages to build (vs. reuse)                      -        0
  2         number of nodes from the same package                        -        0
  3         requirement weight                                           0        0
  4         deprecated versions used                                     0        0
  5         version weight                                               5        0
  6         number of non-default variants (roots)                       0        0
  7         preferred providers for roots                                0        0
  8         default values of variants not being used (roots)            0        0
  9         number of non-default variants (non-roots)                   0        0
  10        preferred providers (non-roots)                              0        0
  11        compiler mismatches that are not from CLI                    0        0
  12        compiler mismatches that are not from CLI                    0        0
  13        non-preferred OS's                                           0        0
  14        version badness                                              5        0
  15        default values of variants not being used (non-roots)        0        0
  16        non-preferred compilers                                      0        0
  17        target mismatches                                            0        0
  18        non-preferred targets                                        0        0
  19        compiler mismatches (runtimes)                               0        0
  20        version badness (runtimes)                                   0        0
  21        non-preferred targets (runtimes)                             0        0
  22        edge wiring                                                  0        0

[+]  zlib-ng@3f35bfccff2d1dacdfe9844712be1e042d028700=2.1.5%apple-clang@15.0.0+compat+new_strategies+opt+pic+shared build_system=autotools arch=darwin-sonoma-m1
[+]      ^gmake@4.4.1%apple-clang@15.0.0~guile build_system=generic arch=darwin-sonoma-m1
[+]      ^gnuconfig@2022-09-17%apple-clang@15.0.0 build_system=generic arch=darwin-sonoma-m1
```

Interested in thoughts? @tgamblin @becker33 @alalazo ?

What are the conditions for re-use we would want to impose on `re-use` for `GitRefVersions` currently? and for the scenario when every branch based version gets converted to a `GitRefVersion` during concretization?

File: lib/spack/spack/solver/asp.py (hunk 1 above)
Comment (psakievich) [reply lib/spack/spack/solver/asp.py]: Yeah good call. Out of curiosity, why do we even have that loop? Are there other cases besides `GitRefVersions` where a user can supply an arbitrary version on the CLI and we expect it to concretize?

Also I wonder if we should not be allowing any arbitrary version at the CLI, but rather require the RHS of the `GitRefVersion` to be known to spack already?

**Comment by psakievich** [comment]
Still thinking about this, but don't have a chance to work on it again yet. My current thought is we need to add a line to the concretizer that does something like: 
```
if prior_install.version is GitRefVersion && spec.version != prior_install.version:
   assign super heigh weight to prior_install
```
The issue is we can't just rely on satisfies b/c we need a `GitRefVersion.satisifies(StandardVersion)` to evaluate to true for the rest of the concretizer logic. Currently we have it so `StandardVersion` can not satisfy `GitRefVersion` as shown [here](https://github.com/spack/spack/issues/38484#issuecomment-2078539611) but that is clearly not sufficient.

**Comment by psakievich** [comment]
Latest push fixes the issue above.  Separate provenance for installs that are standard versions vs git versions allows the concretizer to prefer standard versions again.  Now reuse won't consider git versions first unless they are the only thing installed, and this implicitly restores the requirement that `GitVersion`s have to match exactly since only `GItVersion`s should be considered for reuse when a `GitVersion` shows up in the DAG.

``` console
$ spack solve --show solutions zlib-ng@git.2.1.5=2.1.4 \
> && spack solve --show solutions zlib-ng@git.2.1.5=2.1 \
> && spack solve --show solutions zlib-ng@git.2.1.4=2.1.5 \
> && spack solve --show solutions zlib-ng@git.2.1.5=2.1.5

 -   zlib-ng@git.2.1.5=2.1.4%apple-clang@15.0.0+compat+new_strategies+opt+pic+shared build_system=autotools arch=darwin-sonoma-m1
[+]      ^gmake@4.4.1%apple-clang@15.0.0~guile build_system=generic arch=darwin-sonoma-m1
[+]      ^gnuconfig@2022-09-17%apple-clang@15.0.0 build_system=generic arch=darwin-sonoma-m1

 -   zlib-ng@git.2.1.5=2.1%apple-clang@15.0.0+compat+new_strategies+opt+pic+shared build_system=autotools arch=darwin-sonoma-m1
[+]      ^gmake@4.4.1%apple-clang@15.0.0~guile build_system=generic arch=darwin-sonoma-m1
[+]      ^gnuconfig@2022-09-17%apple-clang@15.0.0 build_system=generic arch=darwin-sonoma-m1

 -   zlib-ng@git.2.1.4=2.1.5%apple-clang@15.0.0+compat+new_strategies+opt+pic+shared build_system=autotools arch=darwin-sonoma-m1
[+]      ^gmake@4.4.1%apple-clang@15.0.0~guile build_system=generic arch=darwin-sonoma-m1
[+]      ^gnuconfig@2022-09-17%apple-clang@15.0.0 build_system=generic arch=darwin-sonoma-m1

[+]  zlib-ng@git.2.1.5=2.1.5%apple-clang@15.0.0+compat+new_strategies+opt+pic+shared build_system=autotools arch=darwin-sonoma-m1
[+]      ^gmake@4.4.1%apple-clang@15.0.0~guile build_system=generic arch=darwin-sonoma-m1
[+]      ^gnuconfig@2022-09-17%apple-clang@15.0.0 build_system=generic arch=darwin-sonoma-m1
```

File: lib/spack/spack/solver/asp.py [hunk 2]
```diff
@@ -115,6 +115,8 @@ class Provenance(enum.IntEnum):
     PACKAGE_PY = enum.auto()
     # An installed spec
     INSTALLED = enum.auto()
```
Comment (psakievich) [inline lib/spack/spack/solver/asp.py:118]: I would like to rename this but it was causing a failure in the unit-tests that I could not uncover.  As far as I can tell it is only used in this file, but maybe there is something in the ASP that is sensitive and not hitting my grep searches? :shrug:

**Comment by psakievich** [comment]
Both of my tests are not working as expected. From a debugger it looks like the installs aren't getting stored in the database so I can't test re-use that way.

**Comment by psakievich** [comment]
Needed an additional mock due to a missing dependency on gcc-runtime in all mock packages. I confirmed that the tests I added fail on develop in the expected manner, and now pass with this PR.

Questions for reviewers:
~~1) Should I swap the package being tested for something more standard?  I need at least 1 variant, and git attributes for these tests~~
2) Should the mock on `runtime` requirements be extended to a more general mock?

**Comment by psakievich** [comment]
@spackbot fix style

**Comment by psakievich** [comment]
Ugh more pain with this:
``` console
$ spack env activate --temp
$ spack config add view:false
$ spack add zlib-ng@git.2.1.5=2.1.5+opt
$ spack add zlib-ng@git.2.1.5=2.1.5~opt
$ spack solve
==> Error: concretization failed for the following reasons:

   1. 'zlib-ng' required multiple values for single-valued variant 'opt'
   2. 'zlib-ng' required multiple values for single-valued variant 'opt'
    Requested '~opt' and '+opt'
        required because zlib-ng@git.2.1.5=2.1.5~opt requested explicitly
        required because zlib-ng@git.2.1.5=2.1.5+opt requested explicitly
s1082338:spack psakiev$ spack config blame concretizer
```

**Comment by psakievich** [comment]
Hmm well looks like that is a separate issue irrespective of the `GitVersion`.

**Comment by jmcarcell** [comment]
I have been building with this PR for a few days since we have many packages that are built from develop and indeed it seems to fix the issue when reconcretizing. I see that sometimes when reconcretizing with the same setup there are a few packages that get rebuilt even though nothing has changed, but that may be unrelated to this PR.

**Comment by psakievich** [comment]
Interesting. If you run into that will you compare the DAG hashes and confirm they are different. There are many things that can cause that behavior. I would not expect this PR to, but it never hurts to check. Feel free to DM me on slack as well.

**Comment by psakievich** [comment]
@spackbot fix style

**Comment by jmcarcell** [comment]
> Interesting. If you run into that will you compare the DAG hashes and confirm they are different. There are many things that can cause that behavior. I would not expect this PR to, but it never hurts to check. Feel free to DM me on slack as well.

I found out that what was happening was that the cxxstd of `xrootd` was being either concretized to 11 or 14 (so I would install and then it would change the cxxstd needing a rebuild of all the dependents of `xrootd` and I remember repeating this process a few times), without changing anything else. Hardcoding the cxxstd of `xrootd` fixed that but I don't have any evidence that it was caused by this PR. After hardcoding cxxstd I haven't found any problems, and we have nightly builds with many packages using `@hash=develop`

File: lib/spack/spack/test/concretize.py [hunk 3]
```diff
@@ -2913,3 +2913,38 @@ def test_spec_filters(specs, include, exclude, expected):
...
+def test_git_ref_version_can_be_reused(
+    install_mockery_mutable_config, do_not_check_runtimes_on_reuse
+):
+    first_spec = spack.spec.Spec("zlib-ng@git.2.1.5=2.1.5").concretized()
+    first_spec.package.do_install(fake=True, explicit=True)
+
+    with spack.config.override("concretizer:reuse", True):
+        second_spec = spack.spec.Spec("zlib-ng@git.2.1.5=2.1.5~opt").concretized()
+        # is_installed(first_spec)
+        assert second_spec.dag_hash() != first_spec.dag_hash()
```
Comment (becker33) [inline lib/spack/spack/test/concretize.py]: This test would be stronger if you use a non-default variant in `first_spec` and leave the variant un-mentioned in `second_spec`

File: lib/spack/spack/test/concretize.py [hunk 4]
```diff
@@ -2913,3 +2913,38 @@ def test_spec_filters(specs, include, exclude, expected):
...
+
+
+@pytest.mark.only_clingo("clingo only re-use feature being tested")
+def test_reuse_prefers_standard_over_git_versions(
```
Comment (becker33) [inline lib/spack/spack/test/concretize.py:3070]: Have you confirmed that this consistently fails on the current develop? I would've thought we want to use a lower version for `standard_spec` to ensure that it would fail without the PR, but if it already consistently picks the wrong one then this is fine.

File: var/spack/repos/builtin.mock/packages/zlib-ng/package.py [hunk 5]
```diff
@@ -0,0 +1,103 @@
+# Copyright 2013-2024 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+from spack.build_systems import autotools, cmake
+from spack.package import *
+
+
+class ZlibNg(AutotoolsPackage, CMakePackage):
```
Comment (becker33) [inline var/spack/repos/builtin.mock/packages/zlib-ng/package.py]: Do we need something as complex as this to make the tests work? It seems like unnecessary overhead, and we don't want to introduce this as a zlib-api provider in the mock repo since so many tests hardcode zlib.

File: lib/spack/spack/solver/asp.py [hunk 6]
```diff
@@ -2046,7 +2048,10 @@ def define_ad_hoc_versions_from_specs(
             # best possible, so they're guaranteed to be used preferentially.
             version = s.versions.concrete
 
-            if version is None or any(v == version for v in self.possible_versions[s.name]):
+            if version is None or (
+                not isinstance(version, vn.GitVersion)
+                and any((v == version) for v in self.possible_versions[s.name])
```
Comment (becker33) [inline lib/spack/spack/solver/asp.py]: Isn't the first constraint redundant since the declared versions can't be GitVersion? Or has another method already stored `GitVersion` constraints in the dict?

**Review by becker33 — CHANGES_REQUESTED** [review]

File: lib/spack/spack/test/concretize.py (hunk 4 above)
Comment (psakievich) [reply lib/spack/spack/test/concretize.py:3070]: I confirmed it failed when I wrote it. I have not tested it since. I didn't discover this failure mode until I fixed the issue with reuse.  Before this PR it would be impossible to expose this behavior due to the bug

https://github.com/spack/spack/pull/43859#issuecomment-2079308154

File: lib/spack/spack/solver/asp.py (hunk 6 above)
Comment (psakievich) [reply lib/spack/spack/solver/asp.py]: Prior installs are in the dictionary as well.

File: var/spack/repos/builtin.mock/packages/zlib-ng/package.py (hunk 5 above)
Comment (psakievich) [reply var/spack/repos/builtin.mock/packages/zlib-ng/package.py]: No not really. I was going for simple and tried to grab a low level package. We need a git attribute, with git ref versions and variants. I originally wrote this with modifying some of the git mock packages but it ended up being invasive on a lot of tests so I just grabbed something else. I can rename it to some dummy package and pair it down.

**Comment by psakievich** [comment]
@spackbot fix style

File: lib/spack/spack/test/concretize.py (hunk 3 above)
Comment (psakievich) [reply lib/spack/spack/test/concretize.py]: Hmmm that is not a reproducer of the bug reported which is why I didn't do that. In this case I would expect it to reuse.
I can add that to this test though.

**Comment by psakievich** [comment]
@spackbot fix style

File: lib/spack/spack/test/concretize.py [hunk 7]
```diff
@@ -2913,3 +2913,45 @@ def test_spec_filters(specs, include, exclude, expected):
...
+        assert third_spec.dag_hash() == first_spec.dag_hash()
+
+
+@pytest.mark.only_clingo("clingo only reuse feature being tested")
+@pytest.mark.parametrize("standard_version", ["2.0.0", "2.1.5", "2.1.6"])
+def test_reuse_prefers_standard_over_git_versions(
+    standard_version, install_mockery_mutable_config, do_not_check_runtimes_on_reuse
+):
+    """
+    order matters in this test. typically reuse would pick the last installed match
```
Comment (becker33) [inline lib/spack/spack/test/concretize.py]:
```suggestion
    order matters in this test. typically reuse would pick the highest versioned installed match
```

**Review by becker33 — CHANGES_REQUESTED** [review]

**Review by becker33 — APPROVED** [review]
